File: ios/Mobile Client/Mobile Client/database/AuditDatabase.swift
//
//  AuditDatabase.swift
//  Mobile Client
//
//  Created by Eric Ruck on 2/21/18.
//  Copyright © 2018 AuditPro. All rights reserved.
//

import Foundation


/**
 * Provides access to our audits that have not yet been uploaded.
 * @author Eric Ruck
 */
class AuditDatabase : BaseDatabase {
	static let DB_NAME = "audit"
	static let DB_VERSION_2_1 = 2
	static let DB_VERSION_2_2 = 3
	static let DB_VERSION_CURRENT = DB_VERSION_2_2

	/**
	 * Instantiates an object to access our stores database.
	 */
	init?() {
		super.init(name: AuditDatabase.DB_NAME, version: AuditDatabase.DB_VERSION_CURRENT)
	}

	/**
	 * Delegates table creation to the subclass.
	 * @return Success flag
	 */
	override func onCreateDatabase() -> Bool {
		return
			AuditTable.createTable(db: self) &&
			ScanTable.createTable(db: self) &&
			ReportTable.createTable(db: self) &&
			NotesTable.createTable(db: self) &&
			ConditionsTable.createTable(db: self)
	}

	/**
	 * Handles table update event from the parent database.
	 * @param lastVersion Last database version of our current tables
	 * @return Success flag
	 */
	override func onUpdateDatabase(lastVersion: Int) -> Bool {
		return
			AuditTable.updateTable(db: self, lastVersion: lastVersion) &&
			ScanTable.updateTable(db: self, lastVersion: lastVersion) &&
			ReportTable.updateTable(db: self, lastVersion: lastVersion) &&
			NotesTable.updateTable(db: self, lastVersion: lastVersion) &&
			ConditionsTable.updateTable(db: self, lastVersion: lastVersion)
	}

	/**
	 * Gets the notes associated with an audit.
	 * @param forAudit Audit whose notes we want
	 * @return Audit notes
	 */
	func getNotes(forAudit: Audit) -> Notes? {
		return self.getNotes(auditId: forAudit.id)
	}

	/**
	 * Gets the notes associated with an audit.
	 * @param auditId Id of audit whose notes we want
	 * @return Audit notes
	 * @throws MobileClientException Database error
	 */
	private func getNotes(auditId: UUID) -> Notes? {
		// Find the record
		return NotesTable.getNotes(db: self, auditId: auditId)
	}

	/**
	 * Updates the notes for an audit.
	 * @param notes Notes to update
	 * @param contents New contents for internal notes
	 * @param store New contents for store notes
	 * @return Error message or nil on success
	 */
	@discardableResult
	func updateNotes(notes: Notes, contents: String, store: String) -> String? {
		return NotesTable.update(db: self, notes: notes, contents: contents, store: store)
	}

	/**
	 * Gets the selected SKU conditions for a product from the current audit.
	 * @param audit Current audit
	 * @param productId Identifies the product
	 * @return The selected conditions or nil if none or error
	 */
	func getSelectedSKUConditions(audit: Audit, productId: Int) -> Set<Int>? {
		return ConditionsTable.getConditions(db: self, audit: audit, productId: productId)
	}

	/**
	 * Updates the selected SKU conditions for a product in the current audit.
	 * @param audit Current audit
	 * @param productId Identifies the product
	 * @param selectedConditions Selected conditions or nil if none
	 * @return Success flag
	 */
	func updateSelectedSKUConditions(audit: Audit, productId: Int, selectedConditions: Set<Int>?) -> Bool {
		// Find the record
		return ConditionsTable.setConditions(db: self, audit: audit, productId: productId, conditions: selectedConditions)
	}

	/**
	 * Gets the number of completed audits waiting for synchronization.
	 * @param userId Current user
	 * @return Count of complete audits for user or nil on error
	 */
	func completeCount(userId: Int) -> Int? {
		return AuditTable.getCount(self, userId: userId, isCompleted: true)
	}

	/**
	 * Gets all of the completed audits waiting for synchronization.
	 * @param userId User identifier
	 * @return List of complete audits or nil on error
	 */
	func getCompleteAudits(userId: Int) -> [Audit]? {
		return AuditTable.getCompleteAudits(self, userId: userId)
	}

	/**
	 * Begins a new audit. Fails if the user already has an audit in progress.
	 * @param userId Current user
	 * @param storeId Audited store
	 * @param storeDescr Store description for UI
	 * @param auditTypeId Type of audit
	 * @param latitude Optional starting latitude
	 * @param longitude Optional starting longitude
	 * @return The audit being started or nil on error
	 */
	func startAudit(userId: Int, storeId: Int, storeDescr: String, auditTypeId: Int,
		latitude: Double?, longitude: Double?) -> Audit? {
		// Make sure there is no audit in progress
		let openCount = AuditTable.getCount(self, userId: userId, isCompleted: false)
		if ((openCount == nil) || (openCount! > 0)) {
			NSLog("There is currently an audit in progress.")
			return nil;
		}

		// Create the new audit
		let start = Audit(userId: userId, storeId: storeId,
			storeDescr: storeDescr, auditTypeId: auditTypeId,
			latitude: latitude, longitude: longitude)
		return AuditTable.insert(self, audit: start) ? start : nil
	}

	/**
	 * Resumes an audit in progress for the user. Returns null if no audit is in progress.
	 * @param userId Current user
	 * @return Audit in progress or nil if none or error
	 */
	func resumeAudit(userId: Int) -> Audit? {
		return AuditTable.getAudit(self, userId: userId)
	}

	/**
	 * Completes the audit in progress.
	 * @param audit Audit in progress
	 * @param latitude Final latitude or null
	 * @param longitude Final longitude or null
	 * @param endTime Optional end of audit time or null
	 * @return Success flag
	 */
	func completeAudit(audit: Audit, latitude: Double?, longitude: Double?, endTime: Date) -> Bool {
		// Update record
		let complete = Audit(source: audit, latitude: latitude, longitude: longitude, endTime: endTime)
		return (complete == nil) ? false : AuditTable.update(self, audit: complete!)
	}

	/**
	 * Reopens a previously closed audit.
	 * @param audit Audit to reopen
	 * @return Reopened audit or nil on error
	 */
	func reopenAudit(_ audit: Audit) -> Audit? {
		// Update record
		let reopen = Audit(reopen: audit)
		return AuditTable.update(self, audit: reopen) ? reopen : nil
	}

	/**
	 * Adds a scan to this audit.
	 * @param scan Scan to add
	 * @return Success flag
	 */
	func addScan(scan: Scan) -> Bool {
		return ScanTable.insert(self, scan: scan)
	}

	/**
	 * Updates a scan in this audit.
	 * @param scan Scan to update
	 * @return Success flag
	 */
	func updateScan(update: Scan) -> Bool {
		return ScanTable.update(self, scan: update)
	}

	/**
	 * Gets a scan from the current audit.
	 * @param audit Current audit
	 * @param productId Identifier of scanned product
	 * @return The scan or nil if none or error
	 */
	func getScan(audit: Audit, productId: Int) -> Scan? {
		return ScanTable.getScan(self, forProductId: productId, inAudit: audit)
	}

	/**
	 * Gets a report from the current audit.
	 * @param audit Current audit
	 * @param productId Identifier of scanned product
	 * @return The report or nil if none or error
	 */
	func getReport(audit: Audit, productId: Int) -> Report? {
		return ReportTable.getReport(self, forProductId: productId, inAudit: audit)
	}

	/**
	 * Adds a report to the audit in progress.
	 * @param report Report to add
	 * @return Success flag
	 */
	func addReport(report: Report) -> Bool {
		return ReportTable.insert(self, report: report)
	}

	/**
	 * Updates a report in the audit in progress.
	 * @param update Report to update
	 * @return Success flag
	 */
	func updateReport(update: Report) -> Bool {
		if (update.id == nil) {
			// The source report is implicit, we actually need to add
			return ReportTable.insert(self, report: update)
		}

		// Update the report
		return ReportTable.update(self, report: update)
	}

	/**
	 * Gets all of the reports for the current audit.
	 * <p>
	 * Some of the reports returned may be implicit, e.g. In Stock for scanned items.  Pass null for
	 * products to return only explicit records.
	 * @param audit Current audit
	 * @param products Products for store, optional, see remarks
	 * @return Audit reports or nil on error
	 */
	func getAllReports(audit: Audit, products: [Product]?) -> [Report]? {
		// Start with the actual reports
		var reports = ReportTable.getReports(self, forAudit: audit);
		if ((reports == nil) || (products == nil)) {
			// We only want the real records
			return reports;
		}

		// Merge with scans
		let scans = ScanTable.getScans(self, forAudit: audit)
		if (scans == nil) {
			return nil
		}
		for scan in scans! {
			var found: Report? = nil
			for seek in reports! {
				if (seek.productId == scan.productId) {
					found = seek
					break;
				}
			}
			if (found == nil) {
				reports!.append(Report(source: scan));
			}
		}

		// Merge with products
		for prod in products! {
			var found: Report? = nil;
			for seek in reports! {
				if (seek.productId == prod.id) {
					found = seek
					break;
				}
			}
			if (found == nil) {
				reports!.append(Report(audit: audit, product: prod));
			}
		}

		// Return all of the reports
		return reports;
	}

	/**
	 * Populates the passed receipt with the out of stock items in the passed audit.
	 * @param receipt Receipt to populate
	 * @param audit Current audit
	 * @param products Products for store
	 * @return Success flag
	 */
	func populateReceipt(receipt: Receipt, audit: Audit, products: [Product]) -> Bool {
		// Start with the actual reports
		guard let reports = ReportTable.getReports(self, forAudit: audit) else {
			return false
		}

		// Sort products for output
		var sortedProducts = [Product](products)
		sortedProducts.sort(by: { $0.productName < $1.productName })

		// Look for all products that will go on the receipt
		let printVoids = Security.optSetting(name: Security.SETTING_PRINT_VOIDS, defaultValue: false)
		let printConditions = Security.optSetting(name: Security.SETTING_PRINT_CONDITIONS, defaultValue: false)
		let printNotes =
			Security.optSetting(name: Security.SETTING_AUDIT_STORE_NOTES, defaultValue: false) &&
			Security.optSetting(name: Security.SETTING_AUDIT_STORE_NOTES, defaultValue: false)
		for product in sortedProducts {
			// Should we include the current product on the receipt?
			for report in reports {
				if (report.productId == product.id) {
					if (report.reorderStatusId == ReorderStatus.OUT_OF_STOCK.id) {
						receipt.addOutOfStockItem(reorderNumber: product.displayReorderCode, productName: product.productName)
					} else if (printVoids && (report.reorderStatusId == ReorderStatus.VOID.id)) {
						receipt.addVoidItem(reorderNumber: product.displayReorderCode, productName: product.productName)
					}
					break;
				}
			}
			if printConditions {
				// Check for SKU conditions
				receipt.addSKUConditions(all: Security.skuConditions,
					from: self.getSelectedSKUConditions(audit: audit, productId: product.id),
					reorderNumber: product.displayReorderCode, productName: product.productName)
			}
		}

		// Should we check store notes?
		if printNotes {
			// Check store notes
			if
				let notes = self.getNotes(forAudit: audit),
				!notes.isStoreEmpty {
					// Add store notes
					receipt.storeNotes = notes.store;
			}
		}
		return true
	}

	/**
	 * Serializes a completed audit to a JSON formatted string.
	 * @param audit Audit to serialize
	 * @return Serialized audit or nil on error
	 */
	func serialize(audit: Audit) -> String? {
		// Get the related scans and reports
		let scans = ScanTable.getJSON(self, forAudit: audit)
		let reports = ReportTable.getJSON(self, forAudit: audit)
		let skuConditions = ConditionsTable.getJSON(db: self, audit: audit)

		// Get our notes
		let notes = getNotes(forAudit: audit)

		// Get the user info
		let user = ([
			"userId": Security.userId,
			"clientId": Security.clientId
		])

		// Generate the result
		let res: [String: Any?] = ([
			"id": audit.id.uuidString,
			"storeId": audit.storeId,
			"auditStartedAt": BaseDatabase.formatDate(audit.auditStartedAt),
			"auditEndedAt": BaseDatabase.formatDate(audit.auditEndedAt),
			"latitudeAtStart": audit.latitudeAtStart,
			"longitudeAtStart": audit.longitudeAtStart,
			"latitudeAtEnd": audit.latitudeAtEnd,
			"longitudeAtEnd": audit.longitudeAtEnd,
			"user": user,
			"scans": scans,
			"reports": reports,
			"notes": notes?.contents,
			"skuConditions": skuConditions,
			"audit_store_note": notes?.store
		])
		let jsonData = try? JSONSerialization.data(withJSONObject: res, options: [])
		return (jsonData == nil) ? nil : String(data: jsonData!, encoding: .utf8)
	}

	/**
	 * Deletes an audit from the local cache.
	 * @param audit Audit to delete
	 * @return Success flag
	 */
	@discardableResult
	func delete(audit: Audit) -> Bool {
		// Start an update transaction
		if (sqlite3_exec(con, "BEGIN TRANSACTION", nil, nil, nil) != SQLITE_OK) {
			logLastError(String(format: "Begin transaction for delete audit %@", audit.id.description))
			return false
		}

		// Execute the delete steps within the transaction
		var operation = String(format: "Delete audit %@ from scan table", audit.id.description)
		var success = ScanTable.delete(self, forAudit: audit)
		if (success) {
			operation = String(format: "Delete audit %@ from report table", audit.id.description)
			success = ReportTable.delete(self, forAudit: audit)
		}
		if (success) {
			operation = String(format: "Delete audit %@ from audit table", audit.id.description)
			success = AuditTable.delete(self, forAudit: audit)
		}
		if (success) {
			operation = String(format: "Delete audit %@ from conditions table", audit.id.description)
			success = ConditionsTable.delete(self, forAudit: audit)
		}

		// Did we succeed?
		if (success) {
			// Yes, commit
			operation = String(format: "Commit transaction for delete audit %@", audit.id.description)
			success = sqlite3_exec(con, "COMMIT TRANSACTION", nil, nil, nil) == SQLITE_OK
		}
		if (!success) {
			// Failed, roll back
			logLastError(operation)
			sqlite3_exec(con, "ROLLBACK TRANSACTION", nil, nil, nil);
		}
		return success;
	}
}

File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/database/AuditDatabase.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.database;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;

import com.auditpro.mobile_client.MobileClientException;
import com.auditpro.mobile_client.entities.Audit;
import com.auditpro.mobile_client.entities.Notes;
import com.auditpro.mobile_client.entities.Product;
import com.auditpro.mobile_client.entities.Receipt;
import com.auditpro.mobile_client.entities.ReorderStatus;
import com.auditpro.mobile_client.entities.Report;
import com.auditpro.mobile_client.entities.Scan;
import com.auditpro.mobile_client.security.Security;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.List;
import java.util.Set;
import java.util.UUID;


/**
 * Provides access to our audits that have not yet been uploaded.
 * @author Eric Ruck
 */
public class AuditDatabase extends BaseDatabase {

	/**
	 * Instantiates an object to access our stores database.
	 * @param context Application context
	 */
	public AuditDatabase(Context context) {
		super(context, "audit", DB_VERSION_CURRENT);
	}

	/**
	 * Delegates table creation to the subclass.
	 * @param db Database in which table needs to be created
	 */
	@Override
	protected void onCreateDb(SQLiteDatabase db) {
		AuditRecord.createTable(db);
		ScanRecord.createTable(db);
		ReportRecord.createTable(db);
		NotesRecord.createTable(db);
		ConditionsRecord.createTable(db);
	}

	/**
	 * Handles table update event from the parent database.
	 * @param db Database in which table needs to be created
	 * @param lastVersion Last database version of our current tables
	 */
	@Override
	protected void onUpdateDb(SQLiteDatabase db, int lastVersion) {
		AuditRecord.updateTable(db, lastVersion);
		ScanRecord.updateTable(db, lastVersion);
		ReportRecord.updateTable(db, lastVersion);
		NotesRecord.updateTable(db, lastVersion);
		ConditionsRecord.updateTable(db, lastVersion);
	}

	/**
	 * Gets the number of completed audits waiting for synchronization.
	 * @param userId Current user
	 * @return Count of complete audits for user
	 * @throws MobileClientException Database error
	 */
	public int completeCount(int userId) throws MobileClientException {
		try {
			return AuditRecord.getCount(getCon(), userId, true);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to count completed audits in database",
					excSql);
		}
	}

	/**
	 * Gets all of the completed audits waiting for synchronization.
	 * @param userId User identifier
	 * @return List of complete audits
	 */
	public List<Audit> getCompleteAudits(int userId) {
		return AuditRecord.getCompleteAudits(getCon(), userId);
	}

	/**
	 * Begins a new audit. Fails if the user already has an audit in progress.
	 * @param userId Current user
	 * @param storeId Audited store
	 * @param storeDescr Store description for UI
	 * @param auditTypeId Type of audit
	 * @param latitude Starting latitude or null
	 * @param longitude Starting longitude or null
	 * @return The audit being started
	 * @throws MobileClientException Audit already in progress (or database error, unlikely)
	 */
	public Audit startAudit(int userId, int storeId, String storeDescr, int auditTypeId,
		Double latitude, Double longitude) throws MobileClientException {
		// Make sure there is no audit in progress
		if (AuditRecord.getCount(getCon(), userId, false) > 0) {
			throw new MobileClientException(
					"There is currently an audit in progress.");
		}
		try {
			// Create the new audit
			AuditRecord start = new AuditRecord(userId, storeId, storeDescr,
					auditTypeId, latitude, longitude);
			start.insertTo(getCon());
			return new Audit(start);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to create new audit in database",
					excSql);
		}
	}

	/**
	 * Resumes an audit in progress for the user. Returns null if no audit is in progress.
	 * @param userId Current user
	 * @return Audit in progress
	 * @throws MobileClientException Likely database error attempting to resume
	 */
	public Audit resumeAudit(int userId) throws MobileClientException {
		try {
			return AuditRecord.getAudit(getCon(), userId);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to locate open audit in database",
					excSql);
		}
	}

	/**
	 * Completes the audit in progress.
	 * @param audit Audit in progress
	 * @param latitute Final latitude or null
	 * @param longitude Final longitude or null
	 * @param endTime Optional end of audit time or null
	 * @throws MobileClientException Database error
	 */
	public void completeAudit(Audit audit, Double latitute, Double longitude, Date endTime)
		throws MobileClientException {
		try {
			// Update record
			AuditRecord complete = new AuditRecord(audit);
			complete.endAudit(latitute, longitude, endTime);
			complete.update(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to update completed audit in database",
					excSql);
		}
	}

	/**
	 * Reopens a previously closed audit
	 * @param audit Audit to reopen
	 * @return Reopened audit
	 * @throws MobileClientException Database error
	 */
	public Audit reopenAudit(Audit audit)
			throws MobileClientException {
		try {
			// Update record
			AuditRecord reopen = new AuditRecord(audit);
			reopen.reopenAudit();
			reopen.update(getCon());
			return new Audit(reopen);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to update completed audit in database",
					excSql);
		}
	}

	/**
	 * Adds a scan to this audit.
	 * @param scan Scan to add
	 * @throws MobileClientException Database error
	 */
	public void addScan(Scan scan) throws MobileClientException {
		try {
			// Create the new record
			new ScanRecord(scan).insert(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to add scan to database",
					excSql);
		}
	}

	/**
	 * Updates a scan in this audit.
	 * @param scan Scan to update
	 * @throws MobileClientException Database error
	 */
	public void updateScan(Scan scan) throws MobileClientException {
		try {
			// Apply the update to the database
			new ScanRecord(scan).update(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to update scan in database",
					excSql);
		}
	}

	/**
	 * Gets a scan from the current audit.
	 * @param audit Current audit
	 * @param productId Identifier of scanned product
	 * @return The scan or null if none
	 * @throws MobileClientException Database error
	 */
	public Scan getScan(Audit audit, int productId) throws MobileClientException {
		try {
			// Find the record
			ScanRecord scan = ScanRecord.getScan(getCon(), audit, productId);
			return (scan == null) ? null : new Scan(scan);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to lookup scan in database",
					excSql);
		}
	}

	/**
	 * Gets a report from the current audit.
	 * @param audit Current audit
	 * @param productId Identifier of scanned product
	 * @return The report or null if none
	 * @throws MobileClientException Database or state error
	 */
	public Report getReport(Audit audit, int productId) throws MobileClientException {
		try {
			// Find the record
			ReportRecord report = ReportRecord.getReport(getCon(), audit, productId);
			return (report == null) ? null : new Report(report);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to lookup report in database",
					excSql);
		}
	}

	/**
	 * Adds a report to the audit in progress.
	 * @param report Report to add
	 * @throws MobileClientException Database or state error
	 */
	public void addReport(Report report) throws MobileClientException {
		try {
			// Create the new report
			ReportRecord rec = new ReportRecord(report);
			rec.insert(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to add report in database",
					excSql);
		}
	}

	/**
	 * Updates a report in the audit in progress.
	 * @param report Report to update
	 * @throws MobileClientException Database or state error
	 */
	public void updateReport(Report report) throws MobileClientException {
		if (report.getId() == null) {
			// The source report is implicit, we actually need to add
			addReport(report);
		}
		try {
			// Create the new report
			ReportRecord rec = new ReportRecord(report);
			rec.update(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to update report in database",
					excSql);
		}
	}

	/**
	 * Gets all of the reports for the current audit.
	 * <p>
	 * Some of the reports returned may be implicit, e.g. In Stock for scanned items.  Pass null for
	 * products to return only explicit records.
	 * @param audit Current audit
	 * @param products Products for store, optional, see remarks
	 * @return Audit reports
	 * @throws MobileClientException Database or state error
	 */
	public List<Report> getAllReports(Audit audit, List<Product> products) throws MobileClientException {
		try {
			// Start with the actual reports
			List<Report> reports = ReportRecord.getReports(getCon(), audit);
			if (products == null) {
				// We only want the real records
				return reports;
			}

			// Merge with scans
			List<ScanRecord> scans = ScanRecord.getScans(getCon(), audit);
			List<Report> append = new ArrayList<>();
			for(ScanRecord scan : scans) {
				Report found = null;
				for(Report seek : reports) {
					if (seek.getProductId() == scan.getProductId()) {
						found = seek;
						break;
					}
				}
				if (found == null) {
					append.add(new Report(scan));
				}
			}
			reports.addAll(append);

			// Merge with products
			append = new ArrayList<>();
			for (Product prod : products) {
				Report found = null;
				for (Report seek : reports) {
					if (seek.getProductId() == prod.getId()) {
						found = seek;
						break;
					}
				}
				if (found == null) {
					append.add(new Report(audit, prod, null));
				}
			}
			reports.addAll(append);
			return reports;
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to get reports for audit in database",
					excSql);
		}
	}

	/**
	 * Populates the passed receipt with the out of stock items in the passed audit.
	 * @param receipt Receipt to populate
	 * @param audit Current audit
	 * @param products Products for store
	 * @throws MobileClientException Database error
	 */
	public void populateReceipt(Receipt receipt, Audit audit, List<Product> products)
			throws MobileClientException {
		try {
			// Start with the actual reports
			List<Report> reports = ReportRecord.getReports(getCon(), audit);

			// Sort products for output
			ArrayList<Product> sortedProducts = new ArrayList<>(products);
			Collections.sort(sortedProducts, new Comparator<Product>() {
				@Override
				public int compare(Product product, Product t1) {
					return product.getProductName().compareTo(t1.getProductName());
				}
			});

			// Look for all products that will go on the receipt
			Security sec = new Security(ctx);
			boolean printVoids = sec.optSettingBool(Security.SETTING_PRINT_VOIDS, false);
			boolean printConditions = sec.optSettingBool(Security.SETTING_PRINT_CONDITIONS, false);
			boolean printNotes =
					sec.optSettingBool(Security.SETTING_AUDIT_STORE_NOTES, false) &&
					sec.optSettingBool(Security.SETTING_PRINT_STORE_NOTES, false);
			for(Product product : sortedProducts) {
				// Should we include the current product on the receipt?
				for (Report report : reports) {
					if (report.getProductId() == product.getId()) {
						if (report.getReorderStatusId() == ReorderStatus.OUT_OF_STOCK.getId()) {
							receipt.addOutOfStockItem(product.getDisplayReorderCode(), product.getProductName());
						} else if (printVoids && (report.getReorderStatusId() == ReorderStatus.VOID.getId())) {
							receipt.addVoidItem(product.getDisplayReorderCode(), product.getProductName());
						}
						break;
					}
				}
				if (printConditions) {
					// Check for SKU conditions
					receipt.addSKUConditions(sec.getSKUConditions(), getSelectedSKUConditions(audit, product.getId()),
							product.getDisplayReorderCode(), product.getProductName());

				}
			}

			// Should we check store notes?
			if (printNotes) {
				// Check store notes
				Notes notes = getNotes(audit);
				if ((notes != null) && !notes.isStoreEmpty()) {
					// Add store notes
					receipt.setStoreNotes(notes.getStore());
				}
			}
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to get reports for audit in database",
					excSql);
		}
	}

	/**
	 * Gets the notes associated with an audit.
	 * @param audit Audit whose notes we want
	 * @return Audit notes
	 * @throws MobileClientException Database error
	 */
	public Notes getNotes(Audit audit) throws MobileClientException {
		return getNotes(audit.getId());
	}

	/**
	 * Gets the notes associated with an audit.
	 * @param auditId Id of audit whose notes we want
	 * @return Audit notes
	 * @throws MobileClientException Database error
	 */
	private Notes getNotes(UUID auditId) throws MobileClientException {
		try {
			// Find the record
			return NotesRecord.getNote(getCon(), auditId);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to lookup notes in database",
					excSql);
		}
	}

	/**
	 * Updates the notes for an audit.
	 * @param notes Notes to update
	 * @param contents New contents for internal notes
	 * @param store New contents for store notes
	 * @throws MobileClientException Database error
	 */
	public void updateNotes(Notes notes, String contents, String store) throws MobileClientException {
		try {
			// Is this a new notes record?
			new NotesRecord(notes, contents, store).update(getCon());
		} catch (SQLiteException excSql) {
			// Unexpected database error
			String method = (notes.getId() == null) ? "create" : "update";
			throw new MobileClientException(
					String.format("Failed to %s notes for audit %s in database",
							method, notes.getAuditId().toString()),
					excSql);
		}
	}

	/**
	 * Gets the selected SKU conditions for a product from the current audit.
	 * @param audit Current audit
	 * @param productId Identifies the product
	 * @return The selected conditions or null if none
	 * @throws MobileClientException Database error
	 */
	public Set<Integer> getSelectedSKUConditions(Audit audit, int productId) throws MobileClientException {
		try {
			// Find the record
			ConditionsRecord conditions = ConditionsRecord.getConditions(getCon(), audit, productId);
			return (conditions == null) ? null : conditions.getConditions();
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to lookup conditions in database",
					excSql);
		}
	}

	/**
	 * Updates the selected SKU conditions for a product in the current audit.
	 * @param audit Current audit
	 * @param productId Identifies the product
	 * @param selectedConditions Selected conditions or null if none
	 * @throws MobileClientException Database error
	 */
	public void updateSelectedSKUConditions(Audit audit, int productId, Set<Integer> selectedConditions)
			throws MobileClientException {
		try {
			// Find the record
			ConditionsRecord.setConditions(getCon(), audit, productId, selectedConditions);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to lookup conditions in database",
					excSql);
		}
	}

	/**
	 * Serialize a completed audit to a JSON formatted string.
	 * @param audit Audit to serialize
	 * @return Serialized audit
	 * @throws MobileClientException Database or serialization fault
	 */
	public String serializeAudit(Audit audit) throws MobileClientException {
		try {
			// Get the related scans and reports
			JSONArray scans = ScanRecord.getJSON(getCon(), audit);
			JSONArray reports = ReportRecord.getJSON(getCon(), audit);
			JSONArray skuConditions = ConditionsRecord.getJSON(getCon(), audit);

			// Get our notes
			Notes notes = getNotes(audit.getId());

			// Get the user info
			JSONObject user = new JSONObject();
			Security sec = new Security(ctx);
			user.put("userId", sec.getUserId());
			user.put("clientId", sec.getClientId());

			// Generate the result
			JSONObject res = new JSONObject();
			res.put("id", audit.getId().toString());
			res.put("storeId", audit.getStoreId());
			res.put("auditStartedAt", BaseDatabase.parseDateTime(audit.getAuditStartedAt()));
			res.put("auditEndedAt", BaseDatabase.parseDateTime(audit.getAuditEndedAt()));
			res.put("latitudeAtStart", audit.getLatitudeAtStart());
			res.put("longitudeAtStart", audit.getLongitudeAtStart());
			res.put("latitudeAtEnd", audit.getLatitudeAtEnd());
			res.put("longitudeAtEnd", audit.getLongitudeAtEnd());
			res.put("user", user);
			res.put("scans", scans);
			res.put("reports", reports);
			res.put("skuConditions", skuConditions);
			res.put("notes", notes.getContents());
			res.put("audit_store_note", notes.getStore());
			return res.toString();
		} catch (JSONException excJSON) {
			// Unexpected sieralization error
			throw new MobileClientException(
					"Failed to serialized details for requested audit",
					excJSON);
		} catch (SQLiteException excSql) {
			// Unexpected database error
			throw new MobileClientException(
					"Failed to get details for requested audit",
					excSql);
		}
	}

	/**
	 * Deletes an audit from the local cache.
	 * @param audit Audit to delete
	 * @throws MobileClientException Database error
	 */
	public void deleteAudit(Audit audit) throws MobileClientException {
		SQLiteDatabase db = getCon();
		try {
			// Delete records related to audit atomically
			db.beginTransaction();
			ScanRecord.deleteFor(getCon(), audit);
			ReportRecord.deleteFor(getCon(), audit);
			AuditRecord.deleteFor(getCon(), audit);
			ConditionsRecord.deleteFor(getCon(), audit);
			db.setTransactionSuccessful();
		} catch (SQLiteException excSql) {
			// Unexpected database error
			db.endTransaction();
			throw new MobileClientException(
					"Failed to remove audit from local cache",
					excSql
			);
		} finally {
			// Complete transaction
			db.endTransaction();
		}
	}

	/**
	 * Gets the current version of this database.
	 * @return Current version code
	 */
	static public int getVersion() {
		return DB_VERSION_CURRENT;
	}

	@SuppressWarnings("unused")
	public static final int DB_VERSION_INIT = 1;
	public static final int DB_VERSION_16 = 2; // Build 16 store notes
	static final int DB_VERSION_18 = 3; // Build 18 SKU conditions
	private static final int DB_VERSION_CURRENT = DB_VERSION_18;
}
